contracts/DAOSignApp.sol

```diff
@@ -0,0 +1,204 @@
+    event NewProofOfAuthority(SignedProofOfAuthority indexed data);
+    event NewProofOfSignature(SignedProofOfSignature indexed data);
+    event NewProofOfAgreement(SignedProofOfAgreement indexed data);
+
+    function getProofOfAuthority(
+        string memory cid
+    ) external view returns (SignedProofOfAuthorityMsg memory);
+
+    function getProofOfSignature(
+        string memory cid
+    ) external view returns (SignedProofOfSignatureMsg memory);
+
+    function getProofOfAgreement(
+        string memory cid
+    ) external view returns (SignedProofOfAgreementMsg memory);
+
+    function storeProofOfAuthority(SignedProofOfAuthority memory data) external;
+
+    function storeProofOfSignature(SignedProofOfSignature memory data) external;
+
+    function storeProofOfAgreement(SignedProofOfAgreement memory data) external;
+}
+
+contract DAOSignApp is DAOSignEIP712, DAOSignBaseApp {
+    mapping(string => SignedProofOfAuthority) poaus;
+    mapping(string => SignedProofOfSignature) posis;
+    mapping(string => SignedProofOfAgreement) poags;
+    mapping(string => address) proof2signer;
+    mapping(string => mapping(address => uint256)) poauSignersIdx;
+
+    constructor() {
+        domain.name = 'daosign';
+        domain.version = '0.1.0';
+        DOMAIN_HASH = hash(domain);
+        initEIP712Types();
+    }
+
+    function memcmp(bytes memory a, bytes memory b) internal pure returns (bool) {
+        return (a.length == b.length) && (keccak256(a) == keccak256(b));
+    }
+
+    function strcmp(string memory a, string memory b) internal pure returns (bool) {
+        return memcmp(bytes(a), bytes(b));
+    }
+
+    function validate(SignedProofOfAuthority memory data) internal pure returns (bool) {
+        require(bytes(data.proofCID).length == IPFS_CID_LENGHT, 'Invalid proof CID');
+        require(strcmp(data.message.app, 'daosign'), 'Invalid struct name');
+        require(strcmp(data.message.name, 'Proof-of-Authority'), 'Invalid struct name');
+        require(bytes(data.message.agreementCID).length == IPFS_CID_LENGHT, 'Invalid CID length');
+        for (uint256 i = 0; i < data.message.signers.length; i++) {
+            require(data.message.signers[i].addr != address(0), 'Invalid signer');
+        }
+        return true;
+    }
+
+    function validate(SignedProofOfSignature memory data) internal view returns (bool) {
+        require(bytes(data.proofCID).length == IPFS_CID_LENGHT, 'Invalid proof CID');
+        require(strcmp(data.message.app, 'daosign'), 'Invalid struct name');
+        require(strcmp(data.message.name, 'Proof-of-Signature'), 'Invalid struct name');
+
+        uint i = poauSignersIdx[data.message.agreementCID][data.message.signer];
+        require(
+            poaus[data.message.agreementCID].message.signers[i].addr == data.message.signer,
+            'Invalid signer'
+        );
+
+        return true;
+    }
+
+    function validate(SignedProofOfAgreement memory data) internal view returns (bool) {
```

Comment:
data.message.timestamp in Proof-of-Agreement should be equal to the last Proof-of-Signature's timestamp. You may want to validate that

Reply:
@ramilexe This is not a true statement. The creation time of the "Proof-of-Agreement" data will never be equal to the time of the last signature